teamcity: failed tests on master: testrace/TestMonotonicInserts, testrace/TestMonotonicInserts/distsql=on, testrace/TestMonotonicInserts/distsql=off

The following tests appear to have failed:

[#239749](https://teamcity.cockroachdb.com/viewLog.html?buildId=239749):

```
--- FAIL: testrace/TestMonotonicInserts (0.000s)
Test ended in panic.

--- FAIL: testrace/TestMonotonicInserts/distsql=on (0.000s)
Race detected!

------- Stdout: -------
E170501 20:24:48.976894 397672 storage/replica_proposal.go:463  [n1,s1,r2/1:/{System/-Max}] could not load SystemConfig span: must retry later due to intent on SystemConfigSpan
E170501 20:24:52.933092 399402 sql/distsqlrun/tablereader.go:195  [client=127.0.0.1:51537,user=root,n2,TableReader=51] scan error: ReadWithinUncertaintyIntervalError: read at time 1493670292.752145100,0 encountered previous write with future timestamp 1493670292.758925467,1 within uncertainty interval
==================
Write at 0x00c42ad47568 by goroutine 506:
  github.com/cockroachdb/cockroach/pkg/roachpb.observedTimestampSlice.update()
      /go/src/github.com/cockroachdb/cockroach/pkg/roachpb/data.go:1298 +0x398
  github.com/cockroachdb/cockroach/pkg/roachpb.(*Transaction).UpdateObservedTimestamp()
      /go/src/github.com/cockroachdb/cockroach/pkg/roachpb/data.go:949 +0x9b
  github.com/cockroachdb/cockroach/pkg/roachpb.(*Transaction).Update()
      /go/src/github.com/cockroachdb/cockroach/pkg/roachpb/data.go:888 +0x42a
  github.com/cockroachdb/cockroach/pkg/kv.(*TxnCoordSender).heartbeat()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender.go:831 +0x8b1
  github.com/cockroachdb/cockroach/pkg/kv.(*TxnCoordSender).heartbeatLoop()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender.go:691 +0x3c0
  github.com/cockroachdb/cockroach/pkg/kv.(*TxnCoordSender).updateState.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender.go:980 +0x8e
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:294 +0x100

Previous read at 0x00c42ad47568 by goroutine 3081:
  github.com/cockroachdb/cockroach/pkg/roachpb.(*Transaction).Update()
      /go/src/github.com/cockroachdb/cockroach/pkg/roachpb/data.go:887 +0x3fb
  github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).send()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:929 +0x998
  github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).sendEndTxnReq()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:545 +0x21a
  github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).commit()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:438 +0x73
  github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).Commit()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:470 +0x4c
  github.com/cockroachdb/cockroach/pkg/sql.commitSQLTransaction()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1369 +0xe4
  github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execStmtInOpenTxn()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1187 +0x2c31
  github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execStmtsInCurrentTxn()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:947 +0xac2
  github.com/cockroachdb/cockroach/pkg/sql.runTxnAttempt()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:855 +0x106
  github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execRequest.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:687 +0x272
  github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).Exec()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:640 +0xe2
  github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execRequest()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:709 +0xb46
  github.com/cockroachdb/cockroach/pkg/sql.(*Executor).ExecuteStatements()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:497 +0x17a
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).executeStatements()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:836 +0xd9
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).handleSimpleQuery()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:504 +0xe1
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).serve()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:432 +0x9f4
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).ServeConn()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:421 +0xe7c
  github.com/cockroachdb/cockroach/pkg/server.(*Server).Start.func9.1()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:619 +0x17e
  github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:142 +0xdc

Goroutine 506 (running) created at:
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:295 +0x172
  github.com/cockroachdb/cockroach/pkg/kv.(*TxnCoordSender).updateState()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender.go:981 +0xc02
  github.com/cockroachdb/cockroach/pkg/kv.(*TxnCoordSender).Send()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender.go:484 +0x5f6
  github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).send()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:528 +0x227
  github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).send()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:862 +0x59a
  github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).(github.com/cockroachdb/cockroach/pkg/internal/client.send)-fm()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:434 +0x96
  github.com/cockroachdb/cockroach/pkg/internal/client.sendAndFill()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:436 +0x13a
  github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).Run()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:434 +0x111
  github.com/cockroachdb/cockroach/pkg/sql.(*tableInserter).finalize()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/tablewriter.go:116 +0x1b0
  github.com/cockroachdb/cockroach/pkg/sql.(*insertNode).Next()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/insert.go:307 +0x170
  github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execClassic()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1435 +0x48b
  github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execStmt()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1563 +0xa63
  github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execStmtInOpenTxn()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1296 +0x5fd
  github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execStmtsInCurrentTxn()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:947 +0xac2
  github.com/cockroachdb/cockroach/pkg/sql.runTxnAttempt()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:855 +0x106
  github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execRequest.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:687 +0x272
  github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).Exec()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:640 +0xe2
  github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execRequest()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:709 +0xb46
  github.com/cockroachdb/cockroach/pkg/sql.(*Executor).ExecuteStatements()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:497 +0x17a
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).executeStatements()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:836 +0xd9
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).handleExecute()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:823 +0x33f
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).serve()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:455 +0xce2
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).ServeConn()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:421 +0xe7c
  github.com/cockroachdb/cockroach/pkg/server.(*Server).Start.func9.1()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:619 +0x17e
  github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:142 +0xdc

Goroutine 3081 (running) created at:
  github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:144 +0x2f8
  github.com/cockroachdb/cockroach/pkg/server.(*Server).Start.func9()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:625 +0x28e
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:215 +0x160
==================
I170501 20:18:45.792271 1 rand.go:76  Random seed: 3597835326531162562
--- FAIL: testrace/TestMonotonicInserts/distsql=off (0.000s)
Test ended in panic.


```
Please assign, take a look and update the issue accordingly.

--- a/pkg/kv/dist_sender.go
+++ b/pkg/kv/dist_sender.go
@@ -530,12 +530,7 @@ func (ds *DistSender) initAndVerifyBatch(
 				//
 				// We already have a clone of our txn (see above), so we can
 				// modify it freely.
-				//
-				// Zero the existing data. That makes sure that if we had
-				// something of size zero but with capacity, we don't re-use the
-				// existing space (which others may also use). This is just to
-				// satisfy paranoia/OCD and not expected to matter in practice.
-				ba.Txn.ResetObservedTimestamps()
+
 				// OrigTimestamp is the HLC timestamp at which the Txn started, so
 				// this effectively means no more uncertainty on this node.
 				ba.Txn.UpdateObservedTimestamp(nDesc.NodeID, ba.Txn.OrigTimestamp)

--- a/pkg/kv/txn_coord_sender.go
+++ b/pkg/kv/txn_coord_sender.go
@@ -548,9 +548,8 @@ func (tc *TxnCoordSender) maybeRejectClientLocked(
 		// educated guess based on the incoming transaction timestamp.
 		return roachpb.NewError(errNoState)
 	case txnMeta.txn.Status == roachpb.ABORTED:
-		txn := txnMeta.txn.Clone()
-		tc.cleanupTxnLocked(ctx, txn)
-		abortedErr := roachpb.NewErrorWithTxn(roachpb.NewTransactionAbortedError(), &txn)
+		tc.cleanupTxnLocked(ctx, txnMeta.txn)
+		abortedErr := roachpb.NewErrorWithTxn(roachpb.NewTransactionAbortedError(), &txnMeta.txn)
 		// TODO(andrei): figure out a UserPriority to use here.
 		newTxn := roachpb.PrepareTransactionForRetry(
 			ctx, abortedErr,
@@ -559,10 +558,9 @@ func (tc *TxnCoordSender) maybeRejectClientLocked(
 		return roachpb.NewError(roachpb.NewHandledRetryableTxnError(
 			abortedErr.Message, txn.ID, newTxn))
 	case txnMeta.txn.Status == roachpb.COMMITTED:
-		txn := txnMeta.txn.Clone()
-		tc.cleanupTxnLocked(ctx, txn)
+		tc.cleanupTxnLocked(ctx, txnMeta.txn)
 		return roachpb.NewErrorWithTxn(roachpb.NewTransactionStatusError(
-			"transaction is already committed"), &txn)
+			"transaction is already committed"), &txnMeta.txn)
 	default:
 		return nil
 	}
@@ -613,8 +611,10 @@ func (tc *TxnCoordSender) cleanupTxnLocked(ctx context.Context, txn roachpb.Tran
 	}
 
 	// The supplied txn may be newer than the one in txnMeta, which is relevant
-	// for stats.
-	txnMeta.txn = txn
+	// for stats. We clone the txn before storing it, as the caller might not
+	// have provided a deep-copy, and we don't want to share Transactions in the
+	// TxnCoordSender's map with anyone.
+	txnMeta.txn = txn.Clone()
 	// Trigger heartbeat shutdown.
 	close(txnMeta.txnEnd)
 	txnMeta.txnEnd = nil

--- a/pkg/roachpb/data.go
+++ b/pkg/roachpb/data.go
@@ -715,7 +715,7 @@ func (t Transaction) LastActive() hlc.Timestamp {
 // keys with the intent spans.
 func (t Transaction) Clone() Transaction {
 	mt := t.ObservedTimestamps
-	if len(mt) != 0 {
+	if mt != nil {
 		t.ObservedTimestamps = make([]ObservedTimestamp, len(mt))
 		copy(t.ObservedTimestamps, mt)
 	}

--- a/pkg/roachpb/errors.go
+++ b/pkg/roachpb/errors.go
@@ -75,6 +75,8 @@ func NewError(err error) *Error {
 }
 
 // NewErrorWithTxn creates an Error from the given error and a transaction.
+//
+// txn is cloned before being stored in Error.
 func NewErrorWithTxn(err error, txn *Transaction) *Error {
 	e := NewError(err)
 	e.SetTxn(txn)

